### DISCUSSION TITLE
INA3221 support?

### DISCUSSION THREAD

**Discussion by gbananachewicz** [arendst/Tasmota#16660] [Q&A]
INA3221 is this module supported ? I found INA226 support only.

**Comment by barbudor** [comment]
Not yet, but how many measures do you need? 
 you can already have up to 4 ina226

**Comment by gbananachewicz** [comment]
3 or 4 measures.
On ina3321 are 3 on one board and it is space saving.( and cheaper).

**Reply by barbudor** [reply]
OK.
I have that in my to-do list for some time
May be I could write the driver this week
Alternatively if you are using ESP32, you may want to try yourself using Berry

**Reply by gbananachewicz** [reply]
Yes i have Esp32 and i will try it. 
Ina3221 can measure Voltage, on ina226 haven't see it. 

I need to measure V i A. 
![image](https://user-images.githubusercontent.com/25414202/192304872-3f9704cf-902b-4b76-8aeb-56ecc420fc79.png)

**Reply by barbudor** [reply]
Yes Ina 226 can read V and I
It's very flexible as you can use either up side or low side shunt resistor

**Reply by pkkrusty** [reply]
I like the ina3221 for a solar battery solution. A current sensor for solar input, battery, and load output, very handy. Maybe also include the voltage divider code in this also, since 26v is pretty low.

**Reply by barbudor** [reply]
As far as I see, Voltage divider would not be possible with the INA3221 as the Vbus+ is common with one of the shunt side (don't remember which one)
That's one of the great point of the INA226 as a very flexible chip

**Reply by gbananachewicz** [reply]
Wheb is the driver ready ?

**Reply by barbudor** [reply]
Sorry, real life priorities...
I should work on it this week end.
Won't take long

**Reply by barbudor** [reply]
![image](https://user-images.githubusercontent.com/2996042/194729904-5a9e0122-c8a4-4b8d-a95d-1f83d2dafeb3.png)
```
00:28:01.522 MQT: tele/nodemcu/SENSOR = {"Time":"2022-10-09T00:28:01+02:00","INA3221":{"Id":"0x40","Voltage":[0.000,0.000,262.080],"Current":[0.000,0.000,0.000],"Power":[0.000,0.000,0.000]}} (retained)
```

Some progress but still some clean-up to do and more testing
Probably Sunday evening should be good

**Reply by barbudor** [reply]
![image](https://user-images.githubusercontent.com/2996042/194770835-05cdb0f2-bddb-40b8-b08d-891802580982.png)

I hope you have your hardware ready 😄 
Test binary based on dev branch : [tasmota-ina3221_1.zip](https://github.com/arendst/Tasmota/files/9742192/tasmota-ina3221_1.zip)
Beware, this version is very talkative, even with log level 2...

Supposed to support up to 4x INA3221 but I've tested only 2 for now
Configuration:
```
Sensor100 <device_number:1..4>, <rshunt1_ohm>, <rshunt2_ohm>, <rshunt3_ohm>
```
For example for a standard module with 0.1 ohm resistors:
```
Sensor100 1, 0.1, 0.1, 0.1
```
Setting the shunt is mandatory otherwise no value is displayed !!!

If you don't want to use a channel, use 0 as the shunt value for this channel

Conifguration is not saved in flash, so you need to use a boot rule to configure at each restart:
```
Rule1 on system#boot do backlog Sensor100 1, 0.1, 0.1, 0.1; Sensor100 2, 0.1, 0.1, 0.1 endon
Rule1 1
```

The Web presentation is 1 channel per line with Voltage, Current and Power which seemed to be the best if you have all 4 devices representing 12 channels
Note that if Theo is not happy with that presentation, I may have to changed to paginated display like Sonoff SPM. We will see.

The SENSOR message is per device as usual with each of Voltage, Current, Power an array of up to 3 values depending on which channels are enabled (based on shunt 0.0)
```
19:27:37.932 MQT: tele/nodemcu/SENSOR = {"Time":"2022-10-09T19:27:37+02:00",
      "INA3221-1":{"Id":"0x40","Voltage":[4.752,0.000,0.000],"Current":[-0.003,0.000,0.000],"Power":[-0.013,0.000,0.000]},
      "INA3221-2":{"Id":"0x41","Voltage":[0.000,0.000,0.000],"Current":[0.000,0.000,0.000],"Power":[0.000,0.000,0.000]}}
```
An not enabled channel (shunt = 0) is represented by `null` in the array in order not to shift value and show that the value is meaningless
An device that is present but which has no active channel is neither displayed nor present in the SENSOR message

PR: https://github.com/arendst/Tasmota/pull/16786

**Reply by gbananachewicz** [reply]
Thank you for your hard work. I have some real world difficulties - some problems to solve. I will test this when I solve priority 1 and 2 (out of 10).  It may take a while. Thanks a lot. Regards Grzegorz

**Reply by gbananachewicz** [reply]
I have 11.54 V

"ANALOG":{"Range":1158},"INA3221":{"Id":"0x40","Voltage":[12,0,0],"Current":[0.00,0.00,0.00],"Power":[0,0,0]}}
 ANALOG RANGE is showing Voltage 11.58v with Resistors.
I need more detailed Voltage no 12V.

must change cable to measure Current (mayby tomorow)
I have R100 resistors on ina3321.  Rule1 on system#boot do backlog Sensor100 1, 100, 0, 0 endon   ?

**Reply by barbudor** [reply]
Voltres AmpRes WattRes are your friends

**Reply by barbudor** [reply]
R100 is 0.1 ohm

**Reply by gbananachewicz** [reply]
Seting 0.1 do not change 12v to other value.

**Reply by sfromis** [reply]
Try `VoltRes 3` (or fewer decimals if you prefer)

**Reply by gbananachewicz** [reply]
THANKS. WORKS> 
"INA3221":{"Id":"0x40","Voltage":[11.656]

**Reply by barbudor** [reply]
My messages are in the order of your questions

**Reply by gbananachewicz** [reply]
Good working.

**Comment by gbananachewicz** [comment]
Working good.

![Screenshot_20221015-140609_copy_720x236](https://user-images.githubusercontent.com/25414202/195985598-6be990a1-810c-4b45-9243-3702d125bf4d.png)

**Reply by sfromis** [reply]
What about setting `WattRes` too? You seem to like decimals. And `AmpRes` for decimals on Current....

**Reply by gbananachewicz** [reply]
Yes. Both are working good.

**Reply by gbananachewicz** [reply]
I see a problem. Ina do not show any Values until Mqtt connect to server. 
@
![Screenshot_20221017-155311_copy_720x392](https://user-images.githubusercontent.com/25414202/196196734-ab228c05-2550-4e27-a983-bce14c625cd1.png)

When will be possible to make own compile with his driver ? Or how?

**Reply by barbudor** [reply]
It is now possible to make your own compile since it was merged
In the above picture the driver is here as the INA3221 is listed but you have not sent the sensor100 command or make the command persistant with a system boot rule

**Reply by gbananachewicz** [reply]
Yes but the rule1 exists and is On.
If I disable mqtt in config then it works normal.

~~~ 
16:27:17.427 RSL: STATE = {"Time":"2022-10-17T16:27:17","Uptime":"0T00:37:26","UptimeSec":2246,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":48,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-35,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:27:17.439 RSL: SENSOR = {"Time":"2022-10-17T16:27:17","ANALOG":{"Range":1148},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"TempUnit":"C"}
16:27:22.516 CMD: Rule1
16:27:22.522 RSL: RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":59,"Free":452,"Rules":"ON system#boot do backlog; Sensor100 1, 0.1, 0.1, 0.1 endon"}}
16:27:27.441 RSL: STATE = {"Time":"2022-10-17T16:27:27","Uptime":"0T00:37:36","UptimeSec":2256,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-36,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:27:27.452 RSL: SENSOR = {"Time":"2022-10-17T16:27:27","ANALOG":{"Range":1148},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"TempUnit":"C"}
16:27:37.444 RSL: STATE = {"Time":"2022-10-17T16:27:37","Uptime":"0T00:37:46","UptimeSec":2266,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-36,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:27:37.456 RSL: SENSOR 
~~~

**Comment by barbudor** [comment]
Please provide logs in both cases as well as the result of 'sensor100 1'
Thanks

**Reply by gbananachewicz** [reply]
Logs please.
MQTT enabled
~~~
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at FB, Count 136
00:00:00.055 QPC: Count 1
00:00:00.068 Project tasmota - Stół Version 12.1.1.4(tasmota)-2_7_4_9(2022-10-09T19:23:54)
00:00:00.088 I2C: INA3221 found at 0x40
00:00:00.527 WIF: Connecting to AP1 axel-gb Channel 11 BSSId E4:6F:13:27:2B:D8 in mode 11n as Wemos3-stol...
00:00:03.754 WIF: Connected
00:00:04.010 HTP: Web server active on Wemos3-stol with IP address 192.168.5.157
16:50:49.551 MQT: Attempting connection...
16:50:49.691 MQT: Connect failed to b37.mqtt.one:1883, rc 5. Retry in 10 sec
16:50:50.575 QPC: Reset
16:50:58.430 RSL: STATE = {"Time":"2022-10-17T16:50:58","Uptime":"0T00:00:16","UptimeSec":16,"Heap":27,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-35,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:50:58.463 RSL: SENSOR = {"Time":"2022-10-17T16:50:58","ANALOG":{"Range":1148},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"TempUnit":"C"}
16:51:01.843 MQT: Attempting connection...
16:51:01.970 MQT: Connect failed to b37.mqtt.one:1883, rc 5. Retry in 20 sec
16:51:08.427 RSL: STATE = {"Time":"2022-10-17T16:51:08","Uptime":"0T00:00:26","UptimeSec":26,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-35,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:51:08.456 RSL: SENSOR = {"Time":"2022-10-17T16:51:08","ANALOG":{"Range":1148},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"TempUnit":"C"}
16:51:18.435 RSL: STATE = {"Time":"2022-10-17T16:51:18","Uptime":"0T00:00:36","UptimeSec":36,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":29,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-36,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:51:18.464 RSL: SENSOR = {"Time":"2022-10-17T16:51:18","ANALOG":{"Range":1146},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"TempUnit":"C"}
16:51:20.901 CMD: Sensor100 1
16:51:20.906 RSL: RESULT = {"Sensor100":{"idx":1,"addr":"0x40","rshunt":[0.00000,0.00000,0.00000]}}
16:51:23.618 MQT: Attempting connection...
16:51:23.809 MQT: Connect failed to b37.mqtt.one:1883, rc 5. Retry in 30 sec
16:51:28.437 RSL: STATE = {"Time":"2022-10-17T16:51:28","Uptime":"0T00:00:46","UptimeSec":46,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":19,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-36,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:51:28.468 RSL: SENSOR = {"Time":"2022-10-17T16:51:28","ANALOG":{"Range":1148},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"TempUnit":"C"}
~~~

Mqtt DISABLED   in Configure Other
~~~
00:00:00.001 HDW: ESP8266EX
00:00:00.050 CFG: Loaded from flash at F5, Count 134
00:00:00.055 QPC: Count 1
00:00:00.068 Project tasmota - Stół Version 12.1.1.4(tasmota)-2_7_4_9(2022-10-09T19:23:54)
00:00:00.088 I2C: INA3221 found at 0x40
00:00:00.527 WIF: Connecting to AP1 axel-gb Channel 11 BSSId E4:6F:13:27:2B:D8 in mode 11n as Wemos3-stol...
00:00:03.757 WIF: Connected
00:00:04.009 HTP: Web server active on Wemos3-stol with IP address 192.168.5.157
00:00:04.013 RSL: INFO1 = {"Info1":{"Module":"Generic","Version":"12.1.1.4(tasmota)","FallbackTopic":"cmnd/2cjlou2394_fb/","GroupTopic":"cmnd/tasmotas/"}}
00:00:04.024 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"Wemos3-stol","IPAddress":"192.168.5.157"}}
00:00:04.033 RSL: INFO3 = {"Info3":{"RestartReason":"Software/System restart","BootCount":64}}
00:00:04.040 RSL: RESULT = {"POWER":"OFF"}
00:00:04.043 RSL: POWER = OFF
00:00:04.068 RUL: SYSTEM#BOOT performs "backlog; Sensor100 1, 0.1, 0.1, 0.1"
00:00:04.110 RSL: RESULT = {"Sensor100":{"idx":1,"addr":"0x40","rshunt":[0.10000,0.10000,0.10000]}}
16:48:57.482 QPC: Reset
16:48:59.826 RSL: STATE = {"Time":"2022-10-17T16:48:59","Uptime":"0T00:00:10","UptimeSec":10,"Heap":25,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":62,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-33,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:48:59.856 RSL: SENSOR = {"Time":"2022-10-17T16:48:59","ANALOG":{"Range":1146},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"INA3221":{"Id":"0x40","Voltage":[11.536,11.528,0.000],"Current":[0.056,0.050,0.000],"Power":[1,1,0]},"TempUnit":"C"}
16:49:09.023 CMD: Sensor100 1
16:49:09.028 RSL: RESULT = {"Sensor100":{"idx":1,"addr":"0x40","rshunt":[0.10000,0.10000,0.10000]}}
16:49:09.450 RSL: STATE = {"Time":"2022-10-17T16:49:09","Uptime":"0T00:00:20","UptimeSec":20,"Heap":26,"SleepMode":"Dynamic","Sleep":50,"LoadAvg":47,"MqttCount":0,"POWER":"OFF","Wifi":{"AP":1,"SSId":"axel-gb","BSSId":"E4:6F:13:27:2B:D8","Channel":11,"Mode":"11n","RSSI":100,"Signal":-33,"LinkCount":1,"Downtime":"0T00:00:05"}}
16:49:09.480 RSL: SENSOR = {"Time":"2022-10-17T16:49:09","ANALOG":{"Range":1148},"DS18B20":{"Id":"0000000C2CDF","Temperature":23.3},"INA3221":{"Id":"0x40","Voltage":[11.528,11.520,0.000],"Current":[0.056,0.062,0.000],"Power":[1,1,0]},"TempUnit":"C"}
~~~

**Reply by barbudor** [reply]
The reason is that your mqtt connection doesn't work
System#boot will not be executed until the connection to the broker is done

1) replace system#boot by system#init in the rule and it will always work
2) solve your mqtt connection problems

If your broker is on the internet you may need to set 'MqttWifiTimeout 1000'

**Reply by gbananachewicz** [reply]
System#init works. 

Thanks, I must compile a version with TLS MQTT to connect to hivemq back after test version for INA3321.
Regards
Grzegorz

**Comment by gbananachewicz** [comment]
Hi i need help in Esp32 have compliled with 
#ifdef USE_INA3221
#undef USE_INA3221
#endif
#define USE_INA3221

Log:
:18:50.222 CMD: Sensor100 1, 0.1, 0.1, 0.1
19:18:50.230 RSL: RESULT = {"Command":"Error"}

**Reply by barbudor** [reply]
How are you compiling ? With which tool ?
Where have you put those lines ?

You made it work some weeks ago, so why do you fail now ?

These are not the recommend lines but they should still work
The recommended way of doing is to add in `user_config_override.h`  these lines:
```
#ifndef USE_INA3221
#define USE_INA3221
#endif
```

**Reply by gbananachewicz** [reply]
On gitpod.io with https://github.com/benzino77/tasmocompiler
I'm compliled it for wemos and it worked, for esp32 want work 
Worked on Wemos D1 mini now I want connect it to Esp32

**Reply by barbudor** [reply]
Something seems wrong with TasmoCompiler @benzino77
I can't choose Development version nor a version more recent than 12.1.1
INA3221 was added in 12.1.1.4

**Reply by barbudor** [reply]
@gbananachewicz How did you compiled the ESP8266 version ?
You may want to try directly using gitpod.io without TasmoCompiler

**Reply by gbananachewicz** [reply]
New workspace from arendst and do not work. https://github.com/arendst/Tasmota
Ina3221 was included in source code ( disabled - I have enabled it )
Ps  I need to compile tasmota32-bluetooth

**Reply by benzino77** [reply]
> Something seems wrong with TasmoCompiler @benzino77 I can't choose Development version nor a version more recent than 12.1.1 INA3221 was added in 12.1.1.4

To be able to choose `Development` branch of Tasmota you need to use `Development` version of TasmoCompiler: https://gitpod.io/#https://github.com/benzino77/tasmocompiler/tree/development
or
docker run --rm --name tasmocompiler -p 3000:3000 benzino77/tasmocompiler:development
See [this ](https://github.com/benzino77/tasmocompiler#development-version) README section.


For more information why Tasmota v12.2.x is not visible on TasmoCompiler side I will direct you to this discussion:
https://github.com/benzino77/tasmocompiler/discussions/301

**Reply by gbananachewicz** [reply]
Works on development version but there were a connection problem with I2C too.

**Reply by benzino77** [reply]
@gbananachewicz 
https://github.com/arendst/Tasmota/blob/c1ea8953cb3f496a5dbf20033a665ad3ad26f136/tasmota/tasmota_xsns_sensor/xsns_100_ina3221.ino#L21

INA3221 requires I2C. Tasmocompiler does not include i2c by default (unless you chose some features from features list which requires i2c). Try with this one in "Custom parameters" field:
```C
#ifdef USE_I2C
  #undef USE_I2C
#endif
#define USE_I2C

#ifdef USE_INA3221
  #undef USE_INA3221
#endif
#define USE_INA3221
```

**Reply by ortegafernando** [reply]
HI, I have also problems compiling for ESP32. I have tried both master and developement trees in GITPOD but no luck.

Uncommenting this line:   #define USE_INA3221  in my_user_config.h

I am still getting:
```
23:46:08.802 CMD: Sensor100
23:46:08.810 MQT: bi/domo/mppt10020/stat/RESULT = {"Command":"Error"}
```

From information webUI:

```
Program Version: 13.2.0.2(tasmota32)
Build Date & Time: 2023-11-11T22:41:01
```

Before I do the same for an ESP8266 (I am updating some hardware in my house) and everything works perfectly, but not with ESP32.

Please, help. Thanks a lot.

**Reply by sfromis** [reply]
First make sure that you even have the sensor and driver available. The output of the command `I2CScan` should have an address between `0x40` and `0x44` included. And driver number 72 should appear in the output of command `I2CDriver`.

When building, it is not the intention that you change *anything*  in `my_user_config.h` but instead make amendments in `user_config_override.h`. However, this should not be a cause for actually missing the driver.

**Reply by ortegafernando** [reply]
Thanks, (about my_user_config.h it was only to make quickly, you are right of course):

```
07:55:08.334 CMD: I2CScan
07:55:08.357 MQT: bi/domo/mppt10020/stat/RESULT = {"I2CScan":"Error 5 at 0x01"}
07:55:31.877 CMD: I2CDriver.
07:55:31.887 MQT: bi/domo/mppt10020/stat/RESULT = {"I2CDriver":"!7,8,9,10,11,12,13,14,15,17,18,20,24,29,31,36,41,42,44,46,48,58,62,65,69,72,76,77,82"}
07:57:12.367 CMD: Sensor100 1, 0.0255, 0.01545, 0.01545
07:57:12.374 MQT: bi/domo/mppt10020/stat/RESULT = {"Command":"Error"}
```

I am using GPIO21 as I2C SDA 1 and GPIO22 as I2C SCL 1

Whit these commands results do you know if it is correctly compiled?  Thanks a lot

**Reply by sfromis** [reply]
I already told you that the address should appear in the output of `I2CScan`, and it clearly does not when you get an error message instead. Forget about the command `Sensor100` until you have I2C working. Errors there are usually about the wiring and electrical setup. With driver 72 included in `I2CDriver` output, you did get it included in the binary.

**Reply by ortegafernando** [reply]
Great, thanks a lot. I am testing so I am using dupon cables between esp32 and esp8266 sockets (NodeMCU) that I have in my breadboard. Let me check wiring. So if 72 driver is included, everything has compiled perfectly. Thanks a lot.

**Comment by buccaneer-jak** [comment]
Hi,
Do any of the compiled binaries for the ESP32C3 include the INA3221 currently?  The command 'I2cDriver72 1' seems to work correctly but unlike when the INA219 is enabled nothing shows in the Main Menu. 
And entering 'Sensor100 1, 0.1, 0.1, 0.05' results in a "Command":"Error"
I have tried flashing all the ESP32C3 binaries with no luck.

**Reply by sfromis** [reply]
None of the precompiled builds (ESP32-C3 or other) includes `#define USE_INA3221`

What you can get from the `I2CDriver` command with or without some driver number is not if it fails or not, but if the output list of driver numbers includes the one you're interested in.

**Reply by buccaneer-jak** [reply]
I was wondering what that list of numbers was!  "I2cDriver14 0" and "I2cDriver14 1" turned on and off the INA219 so I never gave it much thought.

**Reply by barbudor** [reply]
While the command `I2Cdriver` works independently if the driver is present or not in the build, the status reported by that command only lists the available drivers.
In a same way, command `Status 4` reports the same as `I2Cdriver`
